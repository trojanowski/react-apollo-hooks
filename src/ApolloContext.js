'use strict';
exports.__esModule = true;
var react_1 = require('react');
var ApolloContext = react_1['default'].createContext(null);
function ApolloProvider(_a) {
  var client = _a.client,
    children = _a.children;
  return react_1['default'].createElement(
    ApolloContext.Provider,
    { value: client },
    children
  );
}
exports.ApolloProvider = ApolloProvider;
function useApolloClient(overrideClient) {
  var client = react_1.useContext(ApolloContext);
  // Ensures that the number of hooks called from one render to another remains
  // constant, despite the Apollo client read from context being swapped for
  // one passed directly as prop.
  if (overrideClient) {
    return overrideClient;
  }
  if (!client) {
    // https://github.com/apollographql/react-apollo/blob/5cb63b3625ce5e4a3d3e4ba132eaec2a38ef5d90/src/component-utils.tsx#L19-L22
    throw new Error(
      'Could not find "client" in the context or passed in as a prop. ' +
        'Wrap the root component in an <ApolloProvider>, or pass an ' +
        'ApolloClient instance in via props.'
    );
  }
  return client;
}
exports.useApolloClient = useApolloClient;
