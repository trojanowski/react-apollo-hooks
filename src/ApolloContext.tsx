import ApolloClient from 'apollo-client';
import React, { ReactElement, ReactNode, useContext } from 'react';

const ApolloContext = React.createContext<null | ApolloClient<any>>(null);

export interface ApolloProviderProps<TCacheShape> {
  readonly children?: ReactNode;
  readonly client: ApolloClient<TCacheShape>;
}

export function ApolloProvider<TCacheShape = any>({
  client,
  children,
}: ApolloProviderProps<TCacheShape>): ReactElement<
  ApolloProviderProps<TCacheShape>
> {
  return (
    <ApolloContext.Provider value={client}>{children}</ApolloContext.Provider>
  );
}

export function useApolloClient<TCache = object>(): ApolloClient<TCache> {
  const client = useContext(ApolloContext);

  if (!client) {
    // https://github.com/apollographql/react-apollo/blob/5cb63b3625ce5e4a3d3e4ba132eaec2a38ef5d90/src/component-utils.tsx#L19-L22
    throw new Error(
      'Could not find "client" in the context or passed in as a prop. ' +
        'Wrap the root component in an <ApolloProvider>, or pass an ' +
        'ApolloClient instance in via props.'
    );
  }
  return client;
}
