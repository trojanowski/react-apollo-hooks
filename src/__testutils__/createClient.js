'use strict';
exports.__esModule = true;
var apollo_cache_inmemory_1 = require('apollo-cache-inmemory');
var apollo_client_1 = require('apollo-client');
var apollo_link_mock_1 = require('apollo-link-mock');
function createClient(_a) {
  var _b = _a === void 0 ? {} : _a,
    link = _b.link,
    _c = _b.mocks,
    mocks = _c === void 0 ? [] : _c,
    _d = _b.addTypename,
    addTypename = _d === void 0 ? true : _d;
  return new apollo_client_1.ApolloClient({
    cache: new apollo_cache_inmemory_1.InMemoryCache({
      addTypename: addTypename,
    }),
    link: link ? link : new apollo_link_mock_1.MockLink(mocks),
  });
}
exports['default'] = createClient;
