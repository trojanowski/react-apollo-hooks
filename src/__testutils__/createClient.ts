import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { MockLink, MockedResponse } from 'apollo-link-mock';

interface CreateClientOptions {
  readonly addTypename?: boolean;

  readonly link?: ApolloLink;
  readonly mocks?: ReadonlyArray<MockedResponse>;
}

export default function createClient({
  link,
  mocks = [],
  addTypename = true,
}: CreateClientOptions = {}) {
  return new ApolloClient({
    cache: new InMemoryCache({ addTypename }),
    link: link ? link : new MockLink(mocks),
  });
}
