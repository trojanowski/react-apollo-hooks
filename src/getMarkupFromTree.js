'use strict';
exports.__esModule = true;
var React = require('react');
var SSRContext_1 = require('./internal/SSRContext');
var utils_1 = require('./utils');
function getMarkupFromTree(_a) {
  var tree = _a.tree,
    onBeforeRender = _a.onBeforeRender,
    renderFunction = _a.renderFunction;
  var ssrManager = SSRContext_1.createSSRManager();
  function process() {
    try {
      if (onBeforeRender) {
        onBeforeRender();
      }
      var html = renderFunction(
        React.createElement(
          SSRContext_1.SSRContext.Provider,
          { value: ssrManager },
          tree
        )
      );
      if (!ssrManager.hasPromises()) {
        return html;
      }
    } catch (e) {
      if (!utils_1.isPromiseLike(e)) {
        throw e;
      }
      ssrManager.register(e);
    }
    return ssrManager.consumeAndAwaitPromises().then(process);
  }
  return Promise.resolve().then(process);
}
exports.getMarkupFromTree = getMarkupFromTree;
