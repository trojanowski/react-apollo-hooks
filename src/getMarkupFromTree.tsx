import * as React from 'react';
import { SSRContext, createSSRManager } from './internal/SSRContext';

export interface GetMarkupFromTreeOptions {
  tree: React.ReactNode;
  renderFunction: (tree: React.ReactElement<object>) => string;
}

export function getMarkupFromTree({
  tree,
  renderFunction,
}: GetMarkupFromTreeOptions): Promise<string> {
  const ssrManager = createSSRManager();

  function process(): string | Promise<string> {
    const html = renderFunction(
      <SSRContext.Provider value={ssrManager}>{tree}</SSRContext.Provider>
    );

    if (!ssrManager.hasPromises()) {
      return html;
    }

    return ssrManager.consumeAndAwaitPromises().then(process);
  }

  return Promise.resolve().then(process);
}
