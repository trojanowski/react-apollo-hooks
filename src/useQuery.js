'use strict';
var __assign =
  (this && this.__assign) ||
  function() {
    __assign =
      Object.assign ||
      function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
          s = arguments[i];
          for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
      };
    return __assign.apply(this, arguments);
  };
exports.__esModule = true;
var apollo_client_1 = require('apollo-client');
var react_1 = require('react');
var ApolloContext_1 = require('./ApolloContext');
var SSRContext_1 = require('./internal/SSRContext');
var actHack_1 = require('./internal/actHack');
var queryCache_1 = require('./queryCache');
var utils_1 = require('./utils');
function useQuery(query, _a) {
  var _b = _a === void 0 ? {} : _a,
    // Hook options
    _c = _b.ssr,
    // Hook options
    ssr = _c === void 0 ? true : _c,
    _d = _b.skip,
    skip = _d === void 0 ? false : _d,
    _e = _b.suspend,
    suspend = _e === void 0 ? false : _e,
    // Watch options
    pollInterval = _b.pollInterval,
    _f = _b.notifyOnNetworkStatusChange,
    notifyOnNetworkStatusChange = _f === void 0 ? false : _f,
    // Apollo client options
    overrideClient = _b.client,
    context = _b.context,
    metadata = _b.metadata,
    variables = _b.variables,
    actualCachePolicy = _b.fetchPolicy,
    errorPolicy = _b.errorPolicy,
    fetchResults = _b.fetchResults;
  var client = ApolloContext_1.useApolloClient(overrideClient);
  var ssrManager = react_1.useContext(SSRContext_1.SSRContext);
  var ssrInUse = ssr && ssrManager;
  // Skips when `skip: true` or SSRContext passed but `ssr: false`
  var shouldSkip = skip || (ssrManager != null && !ssr);
  var fetchPolicy =
    ssrInUse &&
    // Taken from https://github.com/apollographql/react-apollo/blob/2d7e48b7d0c26e792e1ed26e98bb84d8fba5bb8a/src/Query.tsx#L167-L169
    (actualCachePolicy === 'network-only' ||
      actualCachePolicy === 'cache-and-network')
      ? 'cache-first'
      : actualCachePolicy;
  var watchQueryOptions = react_1.useMemo(
    function() {
      return utils_1.compact({
        context: context,
        errorPolicy: errorPolicy,
        fetchPolicy: fetchPolicy,
        fetchResults: fetchResults,
        metadata: metadata,
        notifyOnNetworkStatusChange: notifyOnNetworkStatusChange,
        pollInterval: pollInterval,
        query: query,
        variables: variables,
      });
    },
    [
      query,
      pollInterval,
      notifyOnNetworkStatusChange,
      context && utils_1.objToKey(context),
      metadata && utils_1.objToKey(metadata),
      variables && utils_1.objToKey(variables),
      fetchPolicy,
      errorPolicy,
      fetchResults,
    ]
  );
  var observableQuery = react_1.useMemo(
    function() {
      return queryCache_1.getCachedObservableQuery(client, watchQueryOptions);
    },
    [client, watchQueryOptions]
  );
  var _g = react_1.useState(0),
    responseId = _g[0],
    setResponseId = _g[1];
  var currentResult = react_1.useMemo(
    function() {
      var result = observableQuery.currentResult();
      return {
        data: result.data,
        error:
          result.errors && result.errors.length > 0
            ? new apollo_client_1.ApolloError({ graphQLErrors: result.errors })
            : result.error,
        errors: result.errors,
        loading: result.loading,
        // don't try to return `networkStatus` when suspense it's used
        // because it's unreliable in that case
        // https://github.com/trojanowski/react-apollo-hooks/pull/68
        networkStatus: suspend ? undefined : result.networkStatus,
        partial: result.partial,
      };
    },
    [shouldSkip, responseId, observableQuery]
  );
  react_1.useEffect(
    function() {
      if (shouldSkip) {
        return;
      }
      var invalidateCurrentResult = function() {
        // A hack to get rid React warnings during tests. The default
        // implementation of `actHack` just invokes the callback immediately.
        // In tests, it's replaced with `act` from react-testing-library.
        // A better solution welcome.
        actHack_1['default'](function() {
          setResponseId(function(x) {
            return x + 1;
          });
        });
      };
      var subscription = observableQuery.subscribe(
        invalidateCurrentResult,
        invalidateCurrentResult
      );
      queryCache_1.invalidateCachedObservableQuery(client, watchQueryOptions);
      return function() {
        subscription.unsubscribe();
      };
    },
    [shouldSkip, observableQuery]
  );
  ensureSupportedFetchPolicy(suspend, fetchPolicy);
  var helpers = {
    fetchMore: observableQuery.fetchMore.bind(observableQuery),
    refetch: observableQuery.refetch.bind(observableQuery),
    startPolling: observableQuery.startPolling.bind(observableQuery),
    stopPolling: observableQuery.stopPolling.bind(observableQuery),
    updateQuery: observableQuery.updateQuery.bind(observableQuery),
  };
  if (shouldSkip) {
    // Taken from https://github.com/apollographql/react-apollo/blob/5cb63b3625ce5e4a3d3e4ba132eaec2a38ef5d90/src/Query.tsx#L376-L381
    return __assign({}, helpers, {
      data: undefined,
      error: undefined,
      loading: false,
      networkStatus: undefined,
    });
  }
  if (currentResult.partial) {
    if (suspend) {
      // throw a promise - use the react suspense to wait until the data is
      // available
      throw observableQuery.result();
    }
    if (ssrInUse) {
      ssrManager.register(observableQuery.result());
    }
  }
  return __assign({}, helpers, currentResult);
}
exports.useQuery = useQuery;
function ensureSupportedFetchPolicy(suspend, fetchPolicy) {
  if (suspend && fetchPolicy && fetchPolicy !== 'cache-first') {
    throw new Error(
      'Fetch policy ' +
        fetchPolicy +
        " is not supported without 'suspend: false'"
    );
  }
}
