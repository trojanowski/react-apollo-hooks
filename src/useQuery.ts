import ApolloClient, {
  ApolloCurrentResult,
  ApolloError,
  ApolloQueryResult,
  FetchMoreOptions,
  FetchMoreQueryOptions,
  NetworkStatus,
  ObservableQuery,
  OperationVariables,
  WatchQueryFetchPolicy,
  WatchQueryOptions,
} from 'apollo-client';
import { DocumentNode } from 'graphql';
import { useContext, useEffect, useMemo, useState } from 'react';
import { useApolloClient } from './ApolloContext';
import { SSRContext } from './internal/SSRContext';
import actHack from './internal/actHack';
import {
  getCachedObservableQuery,
  invalidateCachedObservableQuery,
} from './queryCache';
import { Omit, compact, objToKey } from './utils';

export interface QueryHookState<TData>
  extends Pick<
    ApolloCurrentResult<undefined | TData>,
    'error' | 'errors' | 'loading' | 'partial'
  > {
  data?: TData;
  // networkStatus is undefined for skipped queries or the ones using suspense
  networkStatus: NetworkStatus | undefined;
}

export interface QueryHookOptions<TVariables, TCache = object>
  extends Omit<WatchQueryOptions<TVariables>, 'query'> {
  // watch query options from apollo client
  notifyOnNetworkStatusChange?: boolean;
  pollInterval?: number;
  // custom options of `useQuery` hook
  client?: ApolloClient<TCache>;
  ssr?: boolean;
  skip?: boolean;
  suspend?: boolean;
}

export interface QueryHookResult<TData, TVariables>
  extends QueryHookState<TData>,
    Pick<
      ObservableQuery<TData, TVariables>,
      'refetch' | 'startPolling' | 'stopPolling' | 'updateQuery'
    > {
  fetchMore<K extends keyof TVariables>(
    fetchMoreOptions: FetchMoreQueryOptions<TVariables, K> &
      FetchMoreOptions<TData, TVariables>
  ): Promise<ApolloQueryResult<TData>>;
}

export function useQuery<
  TData = any,
  TVariables = OperationVariables,
  TCache = object
>(
  query: DocumentNode,
  {
    // Hook options
    ssr = true,
    skip = false,
    suspend = false,

    // Watch options
    pollInterval,
    notifyOnNetworkStatusChange = false,

    // Apollo client options
    client: overrideClient,
    context,
    metadata,
    variables,
    fetchPolicy: actualCachePolicy,
    errorPolicy,
    fetchResults,
  }: QueryHookOptions<TVariables, TCache> = {}
): QueryHookResult<TData, TVariables> {
  const client = useApolloClient(overrideClient);
  const ssrManager = useContext(SSRContext);
  const ssrInUse = ssr && ssrManager;

  // Skips when `skip: true` or SSRContext passed but `ssr: false`
  const shouldSkip = skip || (ssrManager != null && !ssr);
  const fetchPolicy =
    ssrInUse &&
    // Taken from https://github.com/apollographql/react-apollo/blob/2d7e48b7d0c26e792e1ed26e98bb84d8fba5bb8a/src/Query.tsx#L167-L169
    (actualCachePolicy === 'network-only' ||
      actualCachePolicy === 'cache-and-network')
      ? 'cache-first'
      : actualCachePolicy;

  const watchQueryOptions: WatchQueryOptions<TVariables> = useMemo(
    () =>
      compact({
        context,
        errorPolicy,
        fetchPolicy,
        fetchResults,
        metadata,
        notifyOnNetworkStatusChange,
        pollInterval,
        query,
        variables,
      }),
    [
      query,

      pollInterval,
      notifyOnNetworkStatusChange,

      context && objToKey(context),
      metadata && objToKey(metadata),
      variables && objToKey(variables),
      fetchPolicy,
      errorPolicy,
      fetchResults,
    ]
  );
  const observableQuery = useMemo(
    () =>
      getCachedObservableQuery<TData, TVariables>(client, watchQueryOptions),
    [client, watchQueryOptions]
  );

  const [responseId, setResponseId] = useState(0);

  const currentResult = useMemo<QueryHookResult<TData, TVariables>>(
    () => {
      const helpers = {
        fetchMore: observableQuery.fetchMore.bind(observableQuery),
        refetch: observableQuery.refetch.bind(observableQuery),
        startPolling: observableQuery.startPolling.bind(observableQuery),
        stopPolling: observableQuery.stopPolling.bind(observableQuery),
        updateQuery: observableQuery.updateQuery.bind(observableQuery),
      };

      const result = observableQuery.currentResult();

      // return the old result data when there is an error
      let data = result.data as TData;
      if (result.error || result.errors) {
        data = {
          ...result.data,
          ...(observableQuery.getLastResult() || {}).data,
        };
      }

      if (shouldSkip) {
        // Taken from https://github.com/apollographql/react-apollo/blob/5cb63b3625ce5e4a3d3e4ba132eaec2a38ef5d90/src/Query.tsx#L376-L381
        return {
          ...helpers,
          data: undefined,
          error: undefined,
          loading: false,
          networkStatus: undefined,
        };
      }

      return {
        ...helpers,
        data,
        error:
          result.errors && result.errors.length > 0
            ? new ApolloError({ graphQLErrors: result.errors })
            : result.error,
        errors: result.errors,
        loading: result.loading,
        // don't try to return `networkStatus` when suspense it's used
        // because it's unreliable in that case
        // https://github.com/trojanowski/react-apollo-hooks/pull/68
        networkStatus: suspend ? undefined : result.networkStatus,
        partial: result.partial,
      };
    },
    [shouldSkip, responseId, observableQuery]
  );

  useEffect(
    () => {
      if (shouldSkip) {
        return;
      }

      const invalidateCurrentResult = () => {
        // A hack to get rid React warnings during tests. The default
        // implementation of `actHack` just invokes the callback immediately.
        // In tests, it's replaced with `act` from react-testing-library.
        // A better solution welcome.
        actHack(() => {
          setResponseId(x => x + 1);
        });
      };
      const subscription = observableQuery.subscribe(
        invalidateCurrentResult,
        invalidateCurrentResult
      );

      invalidateCachedObservableQuery(client, watchQueryOptions);

      return () => {
        subscription.unsubscribe();
      };
    },
    [shouldSkip, observableQuery]
  );

  ensureSupportedFetchPolicy(suspend, fetchPolicy);

  if (currentResult.partial) {
    if (suspend) {
      // throw a promise - use the react suspense to wait until the data is
      // available
      throw observableQuery.result();
    }

    if (ssrInUse) {
      ssrManager!.register(observableQuery.result());
    }
  }

  return currentResult;
}

function ensureSupportedFetchPolicy(
  suspend: boolean,
  fetchPolicy?: WatchQueryFetchPolicy
) {
  if (suspend && fetchPolicy && fetchPolicy !== 'cache-first') {
    throw new Error(
      `Fetch policy ${fetchPolicy} is not supported without 'suspend: false'`
    );
  }
}
