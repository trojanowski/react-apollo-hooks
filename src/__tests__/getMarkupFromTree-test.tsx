import ApolloClient from 'apollo-client';
import { ApolloLink, Observable } from 'apollo-link';
import { MockedResponse } from 'apollo-link-mock';
import gql from 'graphql-tag';
import * as React from 'react';
import { renderToString } from 'react-dom/server';

import { ApolloProvider } from '../ApolloContext';
import createClient from '../__testutils__/createClient';
import { getMarkupFromTree } from '../getMarkupFromTree';
import { QueryHookOptions, useQuery } from '../useQuery';

const AUTH_QUERY = gql`
  {
    isAuthorized
  }
`;

interface AuthQueryResult {
  isAuthorized: boolean;
}

const USER_QUERY = gql`
  {
    currentUser {
      firstName
    }
  }
`;

interface UserQueryResult {
  currentUser: { firstName: string };
}

const MOCKS: MockedResponse[] = [
  {
    request: { query: AUTH_QUERY },
    result: { data: { isAuthorized: true } },
  },
  {
    request: { query: USER_QUERY },
    result: { data: { currentUser: { firstName: 'James' } } },
  },
];

const linkReturningError = new ApolloLink(
  () =>
    new Observable(observer => {
      observer.error(new Error('Simulating network error'));
    })
);

function createMockClient(link?: ApolloLink) {
  return createClient({ link, mocks: MOCKS, addTypename: false });
}

function useAuthDetails(options?: QueryHookOptions<{}>) {
  const { data, loading } = useQuery<AuthQueryResult>(AUTH_QUERY, {
    ...options,
    suspend: false,
  });

  return Boolean(!loading && data && data.isAuthorized);
}

function UserDetails(props: QueryHookOptions<{}>) {
  const { data, loading } = useQuery<UserQueryResult>(USER_QUERY, {
    ...props,
    suspend: false,
  });

  return (
    <div>
      {loading || !data || !data.currentUser
        ? 'Loading user details'
        : data.currentUser.firstName}
    </div>
  );
}

interface UserWrapperProps extends QueryHookOptions<{}> {
  readonly client: ApolloClient<object>;
}

function UserDetailsWrapper({ client, ...props }: UserWrapperProps) {
  return (
    <ApolloProvider client={client}>
      <UserDetails {...props} />
    </ApolloProvider>
  );
}

it('should run through all of the queries that want SSR', async () => {
  const client = createMockClient();

  await expect(
    getMarkupFromTree({
      renderFunction: renderToString,
      tree: <UserDetailsWrapper client={client} />,
    })
  ).resolves.toMatchInlineSnapshot(`"<div>James</div>"`);
});

it('should allow network-only fetchPolicy as an option and still render prefetched data', () => {
  const client = createMockClient();

  return expect(
    getMarkupFromTree({
      renderFunction: renderToString,
      tree: <UserDetailsWrapper client={client} fetchPolicy="network-only" />,
    })
  ).resolves.toMatchInlineSnapshot(`"<div>James</div>"`);
});

it('should allow cache-and-network fetchPolicy as an option and still render prefetched data', () => {
  const client = createMockClient();

  return expect(
    getMarkupFromTree({
      renderFunction: renderToString,
      tree: (
        <UserDetailsWrapper client={client} fetchPolicy="cache-and-network" />
      ),
    })
  ).resolves.toMatchInlineSnapshot(`"<div>James</div>"`);
});

it('should pick up queries deep in the render tree', () => {
  const client = createMockClient();

  const Container = () => (
    <div>
      <span>Hi</span>
      <div>
        <UserDetailsWrapper client={client} fetchPolicy="cache-and-network" />
      </div>
    </div>
  );

  return expect(
    getMarkupFromTree({ renderFunction: renderToString, tree: <Container /> })
  ).resolves.toMatchInlineSnapshot(
    `"<div><span>Hi</span><div><div>James</div></div></div>"`
  );
});

it('should handle nested queries that depend on each other', () => {
  const client = createMockClient();

  const AuthorizedUser = () => {
    const authorized = useAuthDetails();

    return (
      <div>
        <div>Authorized: {String(authorized)}</div>

        <UserDetails skip={!authorized} />
      </div>
    );
  };

  const Container = () => {
    return (
      <ApolloProvider client={client}>
        <AuthorizedUser />
      </ApolloProvider>
    );
  };

  return expect(
    getMarkupFromTree({ renderFunction: renderToString, tree: <Container /> })
  ).resolves.toMatchInlineSnapshot(
    `"<div><div>Authorized: <!-- -->true</div><div>James</div></div>"`
  );
});

it('should return the first of multiple errors thrown by nested wrapped components', () => {
  const client = createMockClient();

  const fooError = new Error('foo');
  const BorkedComponent = () => {
    throw fooError;
  };

  const Container = () => {
    return (
      <div>
        <UserDetailsWrapper client={client} />
        <BorkedComponent />
        <BorkedComponent />
      </div>
    );
  };

  return expect(
    getMarkupFromTree({ renderFunction: renderToString, tree: <Container /> })
  ).rejects.toBe(fooError);
});

it('should handle errors thrown by queries', async () => {
  const client = createMockClient(linkReturningError);
  const tree = <UserDetailsWrapper client={client} />;

  await expect(
    getMarkupFromTree({ tree, renderFunction: renderToString })
  ).rejects.toMatchInlineSnapshot(
    `[Error: Network error: Simulating network error]`
  );

  expect(renderToString(tree)).toMatchInlineSnapshot(
    `"<div>Loading user details</div>"`
  );
});

it('should correctly skip queries', async () => {
  const client = createMockClient();

  await expect(
    getMarkupFromTree({
      renderFunction: renderToString,
      tree: <UserDetailsWrapper client={client} skip />,
    })
  ).resolves.toMatchInlineSnapshot(`"<div>Loading user details</div>"`);

  expect(client.cache.extract()).toEqual({});
});

it('should use the correct default props for a query', async () => {
  const client = createMockClient();

  await getMarkupFromTree({
    renderFunction: renderToString,
    tree: <UserDetailsWrapper client={client} />,
  });

  expect(client.cache.extract()).toMatchInlineSnapshot(`
Object {
  "$ROOT_QUERY.currentUser": Object {
    "firstName": "James",
  },
  "ROOT_QUERY": Object {
    "currentUser": Object {
      "generated": true,
      "id": "$ROOT_QUERY.currentUser",
      "type": "id",
      "typename": undefined,
    },
  },
}
`);
});

it("shouldn't run queries if ssr is turned to off", async () => {
  const client = createMockClient();

  await expect(
    getMarkupFromTree({
      renderFunction: renderToString,
      tree: <UserDetailsWrapper client={client} ssr={false} />,
    })
  ).resolves.toMatchInlineSnapshot(`"<div>Loading user details</div>"`);

  expect(client.cache.extract()).toEqual({});
});

it('should not require `ApolloProvider` to be the root component', () => {
  const client = createMockClient();

  const Root = (props: { children: React.ReactNode }) => <div {...props} />;

  return expect(
    getMarkupFromTree({
      renderFunction: renderToString,
      tree: (
        <Root>
          <UserDetailsWrapper client={client} />
        </Root>
      ),
    })
  ).resolves.toMatchInlineSnapshot(`"<div><div>James</div></div>"`);
});
