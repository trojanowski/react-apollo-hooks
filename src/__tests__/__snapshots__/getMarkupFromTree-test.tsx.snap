// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow cache-and-network fetchPolicy as an option and still render prefetched data 1`] = `"<div>James</div>"`;

exports[`should allow network-only fetchPolicy as an option and still render prefetched data 1`] = `"<div>James</div>"`;

exports[`should correctly skip queries 1`] = `"<div>Loading user details</div>"`;

exports[`should handle errors thrown by queries 1`] = `[Error: Network error: Simulating network error]`;

exports[`should handle errors thrown by queries 2`] = `"<div>Loading user details</div>"`;

exports[`should handle nested queries that depend on each other 1`] = `"<div><div>Authorized: <!-- -->true</div><div>James</div></div>"`;

exports[`should not require \`ApolloProvider\` to be the root component 1`] = `"<div><div>James</div></div>"`;

exports[`should pick up queries deep in the render tree 1`] = `"<div><span>Hi</span><div><div>James</div></div></div>"`;

exports[`should run through all of the queries that want SSR 1`] = `"<div>James</div>"`;

exports[`should use the correct default props for a query 1`] = `
Object {
  "$ROOT_QUERY.currentUser": Object {
    "firstName": "James",
  },
  "ROOT_QUERY": Object {
    "currentUser": Object {
      "generated": true,
      "id": "$ROOT_QUERY.currentUser",
      "type": "id",
      "typename": undefined,
    },
  },
}
`;

exports[`shouldn't run queries if ssr is turned to off 1`] = `"<div>Loading user details</div>"`;
